### ISSUE TITLE
Excempting React Lifecycle methods from `class-methods-use-this` rule

### ISSUE THREAD

**Issue by antosan** [peerigon/eslint-config-peerigon#12]
React requires its lifecycle methods to be instance methods (even if they do not use the `this` keyword). I am running into eslint errors with `"eslint-config-peerigon": "^10.1.1"` because of the [`class-methods-use-this`](https://github.com/peerigon/eslint-config-peerigon/blob/master/base.js#L40) rule which requires that all class methods that do not use  `this`, be made static functions.

I propose to add all react lifecycle methods to the `exceptMethods` option of this rule to ignore these warnings on them.

**Comment by flootr** [comment]
I agree, we should add the React lifecycle methods to the `exceptMethods` array.

Though, I wonder if it is possible to do this only in React context - for example, if we use a linter config for React (do we have one?) and only disable it there so that the rule only applies if we are using it?

**Comment by jhnns** [comment]
Yeah, I agree. Can you list all the lifecycle methods that are allowed to have no `this`?

Probably all methods where you return something, like `getDefaultProps`. However, I would still insist on `this` in lifecycle methods that do not return something, like `componentWillReceiveProps`, because these methods should ultimately lead to a `setState()`. If it's missing, it's a sign for a global side-effect which should usually be avoided. For the rare cases, where the global side-effect is necessary, you can still manually ignore this rule.

Or am I missing something?

**Comment by antosan** [comment]
We could adopt the [airbnb approach](https://github.com/airbnb/javascript/blob/8468ed842314a5d66816927ba6c35f018035cffc/packages/eslint-config-airbnb/rules/react.js#L22-L33) and exempt the following react lifecycle names:

```jsx
  "rules": {
    "class-methods-use-this": ["error", {
      "exceptMethods": [
        "render",
        "getInitialState",
        "getDefaultProps",
        "getChildContext",
        "componentWillMount",
        "componentDidMount",
        "componentWillReceiveProps",
        "shouldComponentUpdate",
        "componentWillUpdate",
        "componentDidUpdate",
        "componentWillUnmount",
      ],
    }],
  }
```

However, as @flootr mentions, we should only apply this override only in React context i.e. [`peerigon/react`](https://github.com/peerigon/eslint-config-peerigon/blob/master/react.js#L11) rules.

If you agree with this I could create a PR.

**Comment by jhnns** [comment]
Personally, I would not include all methods that start with "component" because I think they *should* contain a `this`. In rare cases, where the side-effect is desired, you can disable this rule. Or am I missing something? Do you know a common use case where there is no `this` inside `component` methods?

However, I would relax the rule to the `warn` level. I started to relax rules that are no style issues or obvious coding errors.

**Comment by jhnns** [comment]
Shipped my suggestion with `11.3.0`. Please re-open if we need more exceptions.
